firmware/recovery.c

```diff
@@ -121,28 +128,60 @@ static void recovery_request(void) {
 	msg_write(MessageType_MessageType_WordRequest, &resp);
 }
 
+static bool is_same_mnemonic(const char *new_mnemonic) {
+	/* The execution time of the following code only depends on the
+	 * (public) input.  This avoids timing attacks.
+	 */
+	char diff = 0;
+	uint32_t i = 0;
+	for (; new_mnemonic[i]; i++) {
+		diff |= (storage.mnemonic[i] - new_mnemonic[i]);
+	}
+	diff |= storage.mnemonic[i];
+	return diff == 0;
+}
+
 /* Called when the last word was entered.
  * Check mnemonic and send success/failure.
  */
 static void recovery_done(void) {
 	uint32_t i;
-	strlcpy(storage.mnemonic, words[0], sizeof(storage.mnemonic));
+	char new_mnemonic[sizeof(storage.mnemonic)] = {0};
+
+	strlcpy(new_mnemonic, words[0], sizeof(new_mnemonic));
 	for (i = 1; i < word_count; i++) {
-		strlcat(storage.mnemonic, " ", sizeof(storage.mnemonic));
-		strlcat(storage.mnemonic, words[i], sizeof(storage.mnemonic));
+		strlcat(new_mnemonic, " ", sizeof(new_mnemonic));
+		strlcat(new_mnemonic, words[i], sizeof(new_mnemonic));
 	}
-	if (!enforce_wordlist || mnemonic_check(storage.mnemonic)) {
-		storage.has_mnemonic = true;
-		if (!enforce_wordlist) {
-			// not enforcing => mark storage as imported
-			storage.has_imported = true;
-			storage.imported = true;
+	if (!enforce_wordlist || mnemonic_check(new_mnemonic)) {
+		// New mnemonic is valid.
+		if (!dry_run) {
+			// Update mnemonic on storage.
+			storage.has_mnemonic = true;
+			strlcpy(storage.mnemonic, new_mnemonic, sizeof(new_mnemonic));
+			if (!enforce_wordlist) {
+				// not enforcing => mark storage as imported
+				storage.has_imported = true;
+				storage.imported = true;
+			}
+			storage_commit();
+			fsm_sendSuccess("Device recovered");
+		} else {
+			// Inform the user about new mnemonic correctness (as well as whether it is the same as the current one).
+			const char *msg = is_same_mnemonic(new_mnemonic) ? "(same as current)" : "(differs from current)";
+			layoutDialog(&bmp_icon_ok, NULL, NULL, NULL, "The mnemonic is", "successfully verified.", msg, NULL, NULL, NULL);
```

Comment:
The message

    The mnemonic is
    successfully verified.
    (differs from current)

could be easily misread.  Maybe not say successfully if the mnemonic differs?